dev/com.ibm.ws.jdbc/src/com/ibm/ws/jdbc/DataSourceService.java

```diff
@@ -849,21 +850,45 @@ protected void modified(ComponentContext context, Map<String, Object> newPropert
 
                     vProps.put(key, value);
                 }
-            } else if (flatPrefix.equals(DSConfig.IDENTIFY_EXCEPTION)) {
-                @SuppressWarnings("unchecked")
-                Map<Integer,DSConfig.IdentifyException> errorMappings = (Map<Integer, IdentifyException>) wProps.computeIfAbsent(DSConfig.IDENTIFY_EXCEPTION, k -> new HashMap<>(3));
-                
-                int id = Integer.valueOf(key.substring(dot1 + 1, dot2)); // get the '0' in identifyException.0.foo
-                IdentifyException errorMapping = errorMappings.computeIfAbsent(id, k -> new IdentifyException());
-                String propertyName = key.substring(dot2 + 1); // get the 'foo' in identifyException.0.foo
-                errorMapping.setProperty(propertyName, value);
-
             } else if (flatPrefix.equals(DSConfig.HERITAGE)) {
                 if (DSConfig.HELPER_CLASS.equals(key) || DSConfig.REPLACE_EXCEPTIONS.equals(key))
                     wProps.put(key, value);
             }
         }
-        
+
+        // identifyException, which is a group of
+        // (identifyException.#.sqlState, identifyException.#.errorCode, identifyException.#.as)
+        // is parsed separately to have a predictable order of precedence when collisions occur
+        Map<Object, String> identifications = null;
+        StringBuilder key = new StringBuilder(31).append("identifyException.");
+        for (int i = 0; i < 1000; i++) { // cardinality is capped at 1000 in metatype
+            key.append(i);
+            int dot2 = key.length();
+            key.append(".as");
+            String as = (String) configProps.get(key.toString());
+            if (as == null)
+                break; // no more nested identifyException elements
+            key.replace(dot2 + 1, dot2 + 9, "sqlState");
+            String sqlState = (String) configProps.get(key.toString());
+            key.replace(dot2 + 1, dot2 + 10, "errorCode");
+            Integer errorCode = (Integer) configProps.get(key.toString());
+            key.delete(18, key.length());
```

Comment:
This all seems like unnecessary manual string manipulation. 
Couldn't this be done by using a formatted string:
```java
String keyFormat = "identifyException.%d.%s"
String key;
for (int i = 0; i < 1000; i++) { // cardinality is capped at 1000 in metatype
  key = String.format(keyFormat, i, "as");
  ...
  key = String.format(keyFormat, i, "sqlState);
  ...
}
```

Would make things more readable, but I'm not sure the performance difference between a StringBuilder and StringFormatter.